Updating branch of golang release

vsoch opened the pull request:
This is an update to #1008 a̶n̶d̶ ̶f̶o̶r̶ ̶t̶h̶e̶ ̶t̶i̶m̶e̶ ̶b̶e̶i̶n̶g̶ ̶j̶u̶s̶t̶ ̶a̶ ̶t̶e̶s̶t̶  to see if a rebase with master here fixes our current testing issues. I had not made any changes to code aside from the Go, so my spidey senses are saying because the PR is old, the testing / CI is failing for niche reasons (possibly using an upstream container with incompatible changes) that might be fixed with a rebase. I'm flagging this for now since 1. it could entirely fail, and 2. I am absolutely terrible (as we have established before) with this git history thing, so I'm sure that is wrong. Safe to ignore for now :)

vsoch commented:
Wow - tests never pass to me! Does anyone see a pig flying?

I'm going to try to be brave now to fix up a few of my last commits, at least squashing them (what I originally wanted to do).

vsoch commented:
! :green_apple: :green_book: :green_circle: :green_heart: :green_salad: :tea: :green_square:

milroy commented on resource/reapi/bindings/go/src/fluxmodule/reapi_module.go:
```diff
@@ -0,0 +1,102 @@
+/* int reapi_module_match_allocate (reapi_module_ctx_t *ctx, bool orelse_reserve,
+   const char *jobspec, const uint64_t jobid,
+   bool *reserved,
+   char **R, int64_t *at, double *ov);*/
```
To conform with the other functions, we should have the full description:
```suggestion
/*! Match a jobspec to the "best" resources and either allocate
 *  orelse reserve them. The best resources are determined by
 *  the selected match policy.
 *
 *  \param ctx       reapi_module_ctx_t context object
 *  \param orelse_reserve
 *                   Boolean: if false, only allocate; otherwise, first try
 *                   to allocate and if that fails, reserve.
 *  \param jobspec   jobspec string.
 *  \param jobid     jobid of the uint64_t type.
 *  \param reserved  Boolean into which to return true if this job has been
 *                   reserved instead of allocated.
 *  \param R         String into which to return the resource set either
 *                   allocated or reserved.
 *  \param at        If allocated, 0 is returned; if reserved, actual time
 *                   at which the job is reserved.
 *  \param ov        Double into which to return performance overhead
 *                   in terms of elapse time needed to complete
 *                   the match operation.
 *  \return          0 on success; -1 on error.
 * int reapi_module_match_allocate (reapi_module_ctx_t *ctx, bool orelse_reserve,
 *   const char *jobspec, const uint64_t jobid,
 *  bool *reserved,
 *  char **R, int64_t *at, double *ov);
 */
```

milroy commented on resource/reapi/bindings/go/src/fluxcli/go.mod:
```diff
@@ -1,3 +1,3 @@
-module github.com/cmisale/flux-sched/resource/hlapi/bindings/go/src/fluxcli
+module github.com/flux-framework/flux-sched/resource/reapi/bindings/go/src/fluxcli
```
This change should be a fixup to "reapi: add Golang CLI bindings" and then squashed.

milroy commented on resource/reapi/bindings/go/src/fluxcli/reapi_cli.go:
```diff
@@ -14,226 +14,260 @@ package fluxcli
 #include "reapi_cli.h"
```
This should be a fixup and squashed into "reapi: add Golang CLI bindings" as well.

milroy commented on resource/reapi/bindings/go/src/fluxmodule/reapi_module.go:
```diff
@@ -12,91 +12,109 @@ package fluxcli
 
```
Following the above recommendations, this change should be a fixup to "reapi: add Golang module bindings."

milroy commented on src/test/docker/focal-golang/Dockerfile:
```diff
@@ -14,11 +14,11 @@ RUN sudo apt-get -qq install -y --no-install-recommends \
 	libyaml-cpp-dev \
```
This change should be a fixup to "ci: add focal-golang to test matrix."

milroy commented on .gitignore:
```diff
@@ -104,3 +104,6 @@ compile_flags.txt
 # Rules to ignore auto-generated test harness scripts
```
Please add a descriptive commit message. This commit might need to broken up- let's discuss strategy.

milroy commented on resource/reapi/bindings/go/src/fluxcli/reapi_cli.go:
```diff
@@ -16,25 +16,25 @@ package fluxcli
 import "C"
```
This is excellent!

milroy commented on resource/reapi/bindings/go/src/fluxmodule/reapi_module_test.go:
```diff
@@ -12,17 +12,17 @@ package fluxmodule
-	jobspec, err := ioutil.ReadFile("/root/flux-sched/t/data/resource/jobspecs/basics/test001.yaml")
-	if ctx == nil {
+	mod := NewReapiModule()
+	jobspec, err := os.ReadFile("/root/flux-sched/t/data/resource/jobspecs/basics/test001.yaml")
```
Interesting, what's the advantage of using `os.ReadFile` here?

vsoch replied:
Nothing especially interesting, in fact it's calling the same thing! It's getting deprecated so this is more of an update. 
![image](https://github.com/flux-framework/flux-sched/assets/814322/b1415daf-62d0-4872-8cb2-e4f0b0eb7165)
https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/io/ioutil/ioutil.go;l=30-38

milroy reviewed (changes requested):
Thank you once again for the cleanup and improvements to the initial PR! Adding the error handling, more idiomatic Go, and the wrapper struct will ensure Fluxion is easier to integrate with Golang-based projects.

There is some restructuring of the commits needed to make each commit a self-contained contribution (which is the Flux Framework recommendation). However, strict obedience to this will squash your contributions in fixup commits. Let's discuss how to do a reorganization that reflects your contributions.

One more thing (probably for me): The K8s writer should have a few test cases with `resource-query`. I can add that commit pretty quickly.

cmisale commented on resource/reapi/bindings/go/src/fluxcli/reapi_cli.go:
```diff
@@ -0,0 +1,239 @@
+
+// func ReapiCliGetNode(ctx *ReapiCtx) string {
+// 	return C.GoString(C.reapi_cli_get_node((*C.struct_reapi_cli_ctx)(ctx)))
+// }
```
absolutely yes

vsoch replied:
This must be an old commit because it's gone in my local version! So done!

vsoch commented on resource/reapi/bindings/go/src/fluxcli/reapi_cli.go:
```diff
@@ -0,0 +1,239 @@
+
+/*! Destroy reapi cli context
+*
+* \param ctx           reapi_cli_ctx_t context object
```
> And the [original REAPI text](https://github.com/flux-framework/flux-sched/blob/5ede1d4f8ef13a8b864017fa4f390847d88ffacc/resource/reapi/bindings/c/reapi_cli.h#L31) needs to be fixed, too.

I'll open an issue for this since I think it's separate from the PR here, and I'm sure if I add it I'll bork up the git history even more.

vsoch replied:
https://github.com/flux-framework/flux-sched/issues/1064

vsoch commented:
okay did some fixups and rebase and updated the message  - I'm sure I messed something up but fingers crossed it's still ok!

milroy commented:
After further investigation into K8s writer testing, it looks like the K8s writer actually isn't used by Fluence or the test framework anymore. @cmisale did I miss something or can we drop the K8s writer changes?

cmisale commented:
As we just discussed on Slack, that seems to be some leftover from old discussions. We agreed we can remove that, as it is not being used at all.

vsoch commented:
I don't know the pieces you are referring to, but if it's something you can show me (and I'll add into my squashed commit) I'd be happy to do that! Otherwise, if you want to do some custom git history fu / or just add to an additional commit, that works for me too. I revert to whatever you think is best @cmisale @milroy.

As for the writer code... get otta here, yer out!! :baseball:

milroy commented:
@vsoch that you can just drop commit 3c98e6461ace93db46dbd82d0c31a32665701082 in an interactive rebase and the rest of the PR will work fine (tested).

This PR will need to be merged after PR #1049 since we'll need to sort out the cmake system for Go.

milroy commented:
Alright, PR #1049 was just merged. @vsoch please rebase when you get a chance so we can see what happens to this PR with the cmake build system.

vsoch commented:
This is the error I'm seeing for the go test configure step (there may be others).

```console
-- flux found in PATH and no specific FLUX_CORE_PREFIX specified, using /usr
  -- Checking for module 'flux-core'
  --   Found flux-core, version 0.53.0-173-g714a33f1d
  -- Checking for module 'flux-hostlist'
  --   Found flux-hostlist, version 0.53.0-173-g714a33f1d
  -- Checking for module 'flux-idset'
  --   Found flux-idset, version 0.53.0-173-g714a33f1d
  -- Checking for module 'flux-optparse'
  --   Found flux-optparse, version 0.53.0-173-g714a33f1d
  -- Checking for module 'flux-pmi'
  --   Found flux-pmi, version 0.53.0-173-g714a33f1d
  -- Checking for module 'flux-schedutil'
  --   Found flux-schedutil, version 0.53.0-173-g714a33f1d
  -- Checking for module 'flux-taskmap'
  --   Found flux-taskmap, version 0.53.0-173-g714a33f1d
  CMake Error at cmake/FindFluxCore.cmake:30 (add_library):
    add_library cannot create ALIAS target "flux::core" because target
    "PkgConfig::FLUX_CORE" is imported but not globally visible.
  Call Stack (most recent call first):
    CMakeLists.txt:65 (find_package)
  
  
  CMake Error at cmake/FindFluxCore.cmake:31 (add_library):
    add_library cannot create ALIAS target "flux::hostlist" because target
    "PkgConfig::FLUX_HOSTLIST" is imported but not globally visible.
  Call Stack (most recent call first):
    CMakeLists.txt:65 (find_package)
  
  
  CMake Error at cmake/FindFluxCore.cmake:32 (add_library):
    add_library cannot create ALIAS target "flux::idset" because target
    "PkgConfig::FLUX_IDSET" is imported but not globally visible.
  Call Stack (most recent call first):
    CMakeLists.txt:65 (find_package)
  
  
  CMake Error at cmake/FindFluxCore.cmake:33 (add_library):
    add_library cannot create ALIAS target "flux::pmi" because target
    "PkgConfig::FLUX_PMI" is imported but not globally visible.
  Call Stack (most recent call first):
    CMakeLists.txt:65 (find_package)
  
  
  CMake Error at cmake/FindFluxCore.cmake:34 (add_library):
    add_library cannot create ALIAS target "flux::optparse" because target
    "PkgConfig::FLUX_OPTPARSE" is imported but not globally visible.
  Call Stack (most recent call first):
    CMakeLists.txt:65 (find_package)
  
  
  CMake Error at cmake/FindFluxCore.cmake:35 (add_library):
    add_library cannot create ALIAS target "flux::schedutil" because target
    "PkgConfig::FLUX_SCHEDUTIL" is imported but not globally visible.
  Call Stack (most recent call first):
    CMakeLists.txt:65 (find_package)
  
  
  CMake Error at cmake/FindFluxCore.cmake:36 (add_library):
    add_library cannot create ALIAS target "flux::taskmap" because target
    "PkgConfig::FLUX_TASKMAP" is imported but not globally visible.
  Call Stack (most recent call first):
    CMakeLists.txt:65 (find_package)
  
  
  CMake Error at CMakeLists.txt:73 (find_package):
    find_package called with invalid argument "3.6...<4.0"
  
  
  -- Checking for module 'yaml-cpp'
  --   Found yaml-cpp, version 0.6.2
  -- Checking for module 'libedit'
  --   Found libedit, version 3.1
  -- Checking for module 'libczmq>=3.0.0'
  --   Found libczmq, version 4.2.0
  -- Checking for module 'hwloc>=1.11.1'
  --   Found hwloc, version 2.1.0
  -- Checking for module 'jansson>=2.10'
  --   Found jansson, version 2.12
  -- Checking for module 'uuid'
  --   Found uuid, version 2.34.0
  -- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.71.0/BoostConfig.cmake (found suitable version "1.71.0", minimum required is "1.50") found components: system filesystem graph regex 
  -- Boost version: 1.71.0
  CMake Error at CMakeLists.txt:122 (message):
    Python module yaml not found or version too low, need 3.10 got
  Call Stack (most recent call first):
    CMakeLists.txt:165 (flux_find_python_module)
  
  
  -- Configuring incomplete, errors occurred!
```

vsoch commented:
ahh I bet you we need to update https://github.com/flux-framework/flux-sched/blob/master/src/test/docker/focal/Dockerfile. @milroy okay if I try to match the current focal container? I can keep it in a separate (new) commit.

vsoch commented:
yeah dawg! We are green again!! :green_heart:

milroy commented:
> yeah dawg! We are green again!! 💚

I expect that's because `make check` didn't run `t9001-golang-basic.t`. We need to add the test here and see what happens: https://github.com/flux-framework/flux-sched/blob/cc781dd319f370cbd5b4d42fea83fa78fbfc1377/t/CMakeLists.txt#L83

vsoch commented:
This is a fun PRr - we've said "see what happens" twice now and I'm enjoying the surprises (genuinely)! Aside from waiting to see, the next step of debugging is actually pretty fun.

milroy commented:
@trws do you have some guidance (or doc links) about adapting the [conditionals in the autotools setup](https://github.com/flux-framework/flux-sched/blob/b1e116011b26d50c991c91fead9f330812c599d0/resource/reapi/bindings/go/src/test/Makefile.am ) to cmake?

vsoch commented:
Here is the error message from the tests - it looks like maybe the go binary "main" isn't building or otherwise not found?

```
 83/86 Test #86: t9001-golang-basic.t ..................***Failed    0.17 sec
  expecting success: 
      ${main} --jgf=${jgf} --jobspec=${jobspec1} > 001.R.out &&
      test_cmp 001.R.out ${exp_dir}/001.R.out
  
  /usr/src/t/t9001-golang-basic.t: 2: eval: ../../resource/reapi/bindings/go/src/test/main: not found
  not ok 1 - match allocate 1 slot: 1 socket: 1 core (pol=default)
  #	
  #	    ${main} --jgf=${jgf} --jobspec=${jobspec1} > 001.R.out &&
  #	    test_cmp 001.R.out ${exp_dir}/001.R.out
  #	
  
  expecting success: 
      ${main} --jgf=${jgf} --jobspec=${jobspec2} > 002.R.out &&
      test_cmp 002.R.out ${exp_dir}/002.R.out
  
  /usr/src/t/t9001-golang-basic.t: 2: eval: ../../resource/reapi/bindings/go/src/test/main: not found
  not ok 2 - match allocate 2 slots: 2 sockets: 5 cores 1 gpu 6 memory
  #	
  #	    ${main} --jgf=${jgf} --jobspec=${jobspec2} > 002.R.out &&
  #	    test_cmp 002.R.out ${exp_dir}/002.R.out
  #	
  
  # failed 2 among 2 test(s)
  1..2
  
  84/86 Test #85: t8001-util-ion-R.t ....................   Passed    2.23 sec
  85/86 Test #84: t7000-shell-datastaging.t .............   Passed    5.08 sec
  86/86 Test #78: t4010-match-conf.t ....................   Passed   12.45 sec
  
  99% tests passed, 1 tests failed out of 86
  
  Total Test time (real) = 124.05 sec
  
  The following tests FAILED:
  	 86 - t9001-golang-basic.t (Failed)
```

vsoch commented:
ohh I see (sorry slow, putting 2 and 2 together) it's this block that isn't happening, so thus no main. https://github.com/flux-framework/flux-sched/blob/b1e116011b26d50c991c91fead9f330812c599d0/resource/reapi/bindings/go/src/test/Makefile.am#L28-L31

vsoch commented:
Would it be something like:

```cmake
if(DEFINED ENV{WITH_GO})
  # do go things
endif()
```

vsoch commented:
ok that didn't work - @trws will wait for your advice!

milroy commented:
I was rushing to post the comment and wasn't clear about the objective. We'll need to keep `resource/reapi/bindings/go/src/test/Makefile.am` as is to maintain autotools functionality, but we'll also need to add and modify various CMakeLists to enable building and running the Golang tests in cmake.

Yeah, I think Tom's input will be helpful here.

vsoch commented:
@trws maybe this is an idea? https://github.com/krumberg/cmake_go/tree/main

trws commented:
Sure, generally (even though the cmake language is objectively awful in almost every way) it has a lot less gotchas than the autotools version.

You probably want one of these forms:

```cmake
# If the variable is defined and one of 1, y, yes or on in some capitalization
if (ENABLE_GO)
endif ()

# expanding the variable explicitly will do basically the same thing with more syntax
if (${ENABLE_GO})
endif()

# if the variable is defined and has any value at all, including falsey values (all except the above)
if(DEFINED ENABLE_GO)
endif()

# If you really actually want to check an environment variable, it would be
if ($ENV{WITH_GO})
endif()

# or for if defined
if (DEFINED ENV{WITH_GO})
endif()
```

Conventionally cmake uses `ENABLE_*` for things autotools used `WITH_*` for, but naming isn't super important.  The main thing that helps is this is actually a conditional, it can check values at configure time and has no funny rules about what can and can't be inside it.  It also has about a hundred operators if you want to test more interesting/specific things: https://cmake.org/cmake/help/latest/command/if.html

trws commented:
For more generally getting golang to build, you could try one of the cmake packages for that, but I would probably use machinery similar to what I put in for building the docs.  Use `add_custom_command` and `add_custom_target` to build up the same build command you currently have in the autotools system.  This is one area where cmake feels more cumbersome, but the main reason is it makes you explicitly say which inputs and outputs you care about so everything can be used by downstream targets.  A quick search around found this as a basic example of mixed go/c++ with cmake: https://stackoverflow.com/questions/67163442/how-to-compile-and-link-go-code-into-c-with-cmake

vsoch commented:
okay I can't build this locally - the git hash for cmake is invalid and it's giving me other errors, so I'm going to go through the slow, painful process of pushing and finding out.

trws commented:
The git hash of cmake?

vsoch commented:
Yeah it's trying to... do something with it:

```console
$ ./configure --prefix=/usr/local
-- VER 947d476f
CMake Error at CMakeLists.txt:15 (project):
  VERSION "947d476f" format invalid.


-- Configuring incomplete, errors occurred!
```

trws commented:
Oh, I need to give that a better error message... That's what happens when you haven't fetched the git tags so there's no version to pull from git, and have neither added a version file, set FLUX_SCHED_VERSION as an env var nor set FLUX_SCHED_VER as a cmake define to override it.  Setting one of those or doing `git fetch --tags upstream` or similar to get the tags should take care of that.

vsoch commented:
Are there instructions anywhere for doing this in the container? First it was python yaml:

```
CMake Error at CMakeLists.txt:122 (message):
  Python module yaml not found or version too low, need 3.10 got
Call Stack (most recent call first):
  CMakeLists.txt:165 (flux_find_python_module)
```
Now I have... many more (after fixing that one)

```console
-- VER 0.28.0-69-gc857c85b
-- Building flux-sched version 0.28.0-69-gc857c85b
CMake Error at cmake/FindFluxCore.cmake:30 (add_library):
  add_library cannot create ALIAS target "flux::core" because target
  "PkgConfig::FLUX_CORE" is imported but not globally visible.
Call Stack (most recent call first):
  CMakeLists.txt:65 (find_package)


CMake Error at cmake/FindFluxCore.cmake:31 (add_library):
  add_library cannot create ALIAS target "flux::hostlist" because target
  "PkgConfig::FLUX_HOSTLIST" is imported but not globally visible.
Call Stack (most recent call first):
  CMakeLists.txt:65 (find_package)


CMake Error at cmake/FindFluxCore.cmake:32 (add_library):
  add_library cannot create ALIAS target "flux::idset" because target
  "PkgConfig::FLUX_IDSET" is imported but not globally visible.
Call Stack (most recent call first):
  CMakeLists.txt:65 (find_package)


CMake Error at cmake/FindFluxCore.cmake:33 (add_library):
  add_library cannot create ALIAS target "flux::pmi" because target
  "PkgConfig::FLUX_PMI" is imported but not globally visible.
Call Stack (most recent call first):
  CMakeLists.txt:65 (find_package)


CMake Error at cmake/FindFluxCore.cmake:34 (add_library):
  add_library cannot create ALIAS target "flux::optparse" because target
  "PkgConfig::FLUX_OPTPARSE" is imported but not globally visible.
Call Stack (most recent call first):
  CMakeLists.txt:65 (find_package)


CMake Error at cmake/FindFluxCore.cmake:35 (add_library):
  add_library cannot create ALIAS target "flux::schedutil" because target
  "PkgConfig::FLUX_SCHEDUTIL" is imported but not globally visible.
Call Stack (most recent call first):
  CMakeLists.txt:65 (find_package)


CMake Error at cmake/FindFluxCore.cmake:36 (add_library):
  add_library cannot create ALIAS target "flux::taskmap" because target
  "PkgConfig::FLUX_TASKMAP" is imported but not globally visible.
Call Stack (most recent call first):
  CMakeLists.txt:65 (find_package)


CMake Error at CMakeLists.txt:73 (find_package):
  find_package called with invalid argument "3.6...<4.0"


-- Boost version: 1.71.0
CMake Error at CMakeLists.txt:122 (message):
  Python module yaml not found or version too low, need 3.10 got
Call Stack (most recent call first):
  CMakeLists.txt:165 (flux_find_python_module)


-- Configuring incomplete, errors occurred!
See also "/workspaces/fs-rebase-two/CMakeFiles/CMakeOutput.log".
```
I'm guessing there are some libraries that aren't on paths that need to be, or something like that?

trws commented:
That almost looks like too low a cmake version maybe? It's like the alias command isn't even working, which is pretty basic necessity stuff. Is this using one of the flux-sched containers or a container with the distro's cmake in it?  Even then it should have blown up at the cmake required version line but that's what it looks like...

vsoch commented:
> That almost looks like too low a cmake version maybe? It's like the alias command isn't even working, which is pretty basic necessity stuff. Is this using one of the flux-sched containers or a container with the distro's cmake in it? Even then it should have blown up at the cmake required version line but that's what it looks like...

Ah good call. I'll update the development environment.

trws commented:
Darn it, I forgot to put up a PR for a devcontainer update, if you swap the devcontainer to use the bookworm one and fix the python-yaml to python3-yaml it should behave itself.

vsoch commented:
yay! That was the issue. Phew. 
![image](https://github.com/flux-framework/flux-sched/assets/814322/1b9aee45-8b17-4b89-b567-3f3ffa86b0f3)

vsoch commented:
okay - I didn't see it in the build log, so I suspect I'm missing an if statement somewhere! 
![image](https://github.com/flux-framework/flux-sched/assets/814322/a386616c-f7ce-4e62-9a5b-9a372a05c760)

It's getting darker earlier and sunset is in ~an hour so I need to go running, but I'll be back after to debug this further.

milroy commented:
> A quick search around found this as a basic example of mixed go/c++ with cmake: https://stackoverflow.com/questions/67163442/how-to-compile-and-link-go-code-into-c-with-cmake

Thanks for the direction and tops with the conditionals! I found that example, too, and am working on an implementation based on it. Updates to follow...

vsoch commented:
And I'll hold off what I'm doing - I'll instead put in a separate PR to update the Development container with jammy, python3-yaml, etc.

vsoch commented:
Updating the PR with (a start) to fixing the cmake build for golang - TBA to see what the next issue is! In summary:

- To build the main test binary we use a custom function in cmake/GolangSimple.cmake. That uses add_custom_command under the hood. The remaining issue there is passing the list of flags in as a variable - when there is a space they always get escaped, which we don't want. I wasn't sure how to pass them properly from the argument into there so any tips would be appreciated!
- A related bug to the above was that the fluxcli was just imported as "fluxcli" but then go was looking for it in GOROOT. Changing GOPATH didn't work. I think generally we should follow the practice of having a GitHub path, so I moved the go.mod one level up. The paths are the same, but now we can discover the module.
- The paths for include/link needed to be changed from a autotools convention (in .libs) to cmake (in the directory where they are built for the most part)
- For testing / building logic, I added checks that the `WITH_GO` envar is set. This could be more explicit (e.g., checking that it's yes/on but I thought this was OK for now. If we want to be more explicit a "check if this equals" example would be helpful (or I guess I could Google it too... lol)
- The hardest bug was that `Flux::Jobspec::Jobspec` was undefined, which first showed up in an error, and then revealed itself to be true using nm on the *.a file. We naively thought this was in flux-core, but when I poked around there I was like derp, that looks like C++ and this is in C! So then I started searching around flux-sched (c++!) and found it in the libjobspec directory. I'm not sure how this built before, but I'm guessing it was built into that a (and thus defined) so maybe a difference due to the change in the build system.
- Make clean works (fairly easily) by adding `main` as a clean file via add property.

I'll add more as I remember - that's all in my head for now! I already see failing checks, will investigate for next steps! I didn't actually run any tests locally, just got the `make` and `make clean` working.

vsoch commented:
Going to need some help on the new failures (for install) I don't see what the issue is, just lots of failing tests :)

vsoch commented:
okay - for sanity check I cloned fresh and ran `make` and  am running `make check` now to see if I can reproduce the issues. After that I'll try `make clean` and build without Go and see if I can reproduce the install issues.

vsoch commented:
My hypothesis for this:

```
-ljobspec_conv -lreapi_cli -L/usr/src/resource -lresource -lflux-idset -lstdc++ -lczmq -ljansson -lhwloc -lboost_system -lflux-hostlist -lboost_graph -lyaml-cpp' go build -ldflags '-w' -o /usr/src/resource/reapi/bindings/go/src/test/main main.go
  # command-line-arguments
  /usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
  /usr/bin/ld: cannot find -ljobspec_conv
  /usr/bin/ld: cannot find -lresource
  /usr/bin/ld: cannot find -ljobspec_conv
  /usr/bin/ld: cannot find -lresource
  collect2: error: ld returned 1 exit status
```
Is that if this is running with a `-j 4` or similar, we have to tell cmake these are dependencies and not to try building this test until they exist (akin to the others I added that don't show up there).

vsoch commented:
okay getting closer - the go is building! This is the current test issue - it's not finding a library. I suspect I need to figure out how exactly that works.

```
85/86 Test #86: t9001-golang-basic.t ..................***Failed    0.24 sec
  expecting success: 
      ${main} --jgf=${jgf} --jobspec=${jobspec1} > 001.R.out &&
      test_cmp 001.R.out ${exp_dir}/001.R.out
  
  ../../resource/reapi/bindings/go/src/test/main: error while loading shared libraries: libresource.so: cannot open shared object file: No such file or directory
  not ok 1 - match allocate 1 slot: 1 socket: 1 core (pol=default)
  #	
  #	    ${main} --jgf=${jgf} --jobspec=${jobspec1} > 001.R.out &&
  #	    test_cmp 001.R.out ${exp_dir}/001.R.out
  #	
  
  expecting success: 
      ${main} --jgf=${jgf} --jobspec=${jobspec2} > 002.R.out &&
      test_cmp 002.R.out ${exp_dir}/002.R.out
  
  ../../resource/reapi/bindings/go/src/test/main: error while loading shared libraries: libresource.so: cannot open shared object file: No such file or directory
  not ok 2 - match allocate 2 slots: 2 sockets: 5 cores 1 gpu 6 memory
  #	
  #	    ${main} --jgf=${jgf} --jobspec=${jobspec2} > 002.R.out &&
  #	    test_cmp 002.R.out ${exp_dir}/002.R.out
  #	
  
  # failed 2 among 2 test(s)
  1..2
  
  86/86 Test #78: t4010-match-conf.t ....................   Passed   14.05 sec
  
  99% tests passed, 1 tests failed out of 86
  
  Total Test time (real) = 140.87 sec
  
  The following tests FAILED:
  	 86 - t9001-golang-basic.t (Failed)
```
And @grondo is on top of the issue with the test-install failing, some stray IFS that needs a little "convincing" :fist_oncoming:

vsoch commented:
Golang basic testing worked for me locally. Need to figure out how to reproduce the above.

```
85/86 Test #85: t8001-util-ion-R.t ....................   Passed    1.12 sec
      Start 86: t9001-golang-basic.t
86/86 Test #86: t9001-golang-basic.t ..................   Passed    0.09 sec

100% tests passed, 0 tests failed out of 86

Total Test time (real) = 337.48 sec
Built target check
```

vsoch commented:
Knowing the hard coded ldflags wouldn't work, I found (what I thought was) a trick to write a file first and then execute it for the command - when written into the script (this commit  https://github.com/researchapps/flux-sched/commit/1aa3127018f08b30ef48fae240b0cb2218595318) the previously working build errors and tells me there is no such file. When I move that one level up into the go/src/test directory I still get the same (and those two commits are now here: https://github.com/researchapps/flux-sched/commit/fd79024f9f0ef37eb80f67ef27026301979d5947). I've reverted this because it's one step back, even if the cmake code looks a little nicer.

So it's building again:

![image](https://github.com/flux-framework/flux-sched/assets/814322/35f7e515-1568-43ca-b4bb-2afad12d154a)

but with the same errors as before (make check, which works locally for me, fails, and there is that bug with IFS and the test-install). That's better than not building, but this isn't a solution that is good because the flags are hard coded, but I did try. Hopefully someone smarter than me can provide guidance. Sorry I'm only a few days into doing cmake -  I am starting to see how it works but largely just suck. I think I'm going to stop tonight because I'm definitely making it worse. :laughing:

vsoch commented:
Two fixed (thanks @grondo !) one more to go.

vsoch commented:
It's green!! Ahh @cmisale @milroy we are almost there!! This one was easier because I could see libresource isn't linked:

```bash
$ ldd resource/reapi/bindings/go/src/test/main
```
```console
        linux-vdso.so.1 (0x00007ffdeb48e000)
        libresource.so => not found
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fece4d41000)
        libjansson.so.4 => /lib/x86_64-linux-gnu/libjansson.so.4 (0x00007fece4d31000)
        libflux-idset.so.1 => /usr/lib/libflux-idset.so.1 (0x00007fece4d2a000)
        libflux-hostlist.so.1 => /usr/lib/libflux-hostlist.so.1 (0x00007fece4d23000)
        libyaml-cpp.so.0.7 => /lib/x86_64-linux-gnu/libyaml-cpp.so.0.7 (0x00007fece4ccd000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fece4cad000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fece4acc000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fece49ed000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fece4f63000)
```
That said, I'm not sure why the tests pass locally, I'll reduce it to something mysterious living inside my container. The solution... link it!  Link it, good :sunglasses: 

```bash
$ ldd resource/reapi/bindings/go/src/test/main
```
```console
        linux-vdso.so.1 (0x00007ffe1ebdd000)
        libresource.so => /workspaces/fs-rebase-3/resource/libresource.so (0x00007fcbb9813000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fcbb95f3000)
        libjansson.so.4 => /lib/x86_64-linux-gnu/libjansson.so.4 (0x00007fcbb95e3000)
        libflux-idset.so.1 => /usr/lib/libflux-idset.so.1 (0x00007fcbb95dc000)
        libflux-hostlist.so.1 => /usr/lib/libflux-hostlist.so.1 (0x00007fcbb95d5000)
        libyaml-cpp.so.0.7 => /lib/x86_64-linux-gnu/libyaml-cpp.so.0.7 (0x00007fcbb957f000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fcbb955f000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fcbb937e000)
        libboost_graph.so.1.74.0 => /lib/x86_64-linux-gnu/libboost_graph.so.1.74.0 (0x00007fcbb933d000)
        libhwloc.so.15 => /lib/x86_64-linux-gnu/libhwloc.so.15 (0x00007fcbb92e0000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fcbb91ff000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fcbb9937000)
        libboost_regex.so.1.74.0 => /lib/x86_64-linux-gnu/libboost_regex.so.1.74.0 (0x00007fcbb90f1000)
        libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007fcbb90c3000)
        libicui18n.so.72 => /lib/x86_64-linux-gnu/libicui18n.so.72 (0x00007fcbb8d99000)
        libicuuc.so.72 => /lib/x86_64-linux-gnu/libicuuc.so.72 (0x00007fcbb8b9b000)
        libicudata.so.72 => /lib/x86_64-linux-gnu/libicudata.so.72 (0x00007fcbb6dc7000)
```
So now we have testing that is fully green! I'm going to take a moment to celebrate! :partying_face: 
And now back to figuring out how to get that ldflags variable to actually be used (without escaped spaces). :space_invader:

vsoch commented:
>  And now back to figuring out how to get that ldflags variable to actually be used (without escaped spaces). 👾

Boum. :boom: And we're done. :fork_and_knife:

vsoch commented:
Going to squash commits now that I know it works...

milroy commented:
Excellent work figuring out the CMake build! You can drop commit 0824c9f4aefae2d8005cdb249aedf35d61f2be16 since that updates the deprecated autotools. We'll be removing the autotools files in a PR soon I hope.

trws commented:
Yeah, I plan to put up a PR for that, with the caveat that I want to see at least one successful release before we actually remove autotools.  Until we've seen it go through buildbot without issues using the cmake process I want us to have the option for anything we deploy as part of the rpms.

milroy commented:
>  Until we've seen it go through buildbot without issues using the cmake process I want us to have the option for anything we deploy as part of the rpms.

Ok, good to know @trws. @vsoch you can ignore my previous comment about dropping the autotools-related commit.

grondo commented:
> Yeah, I plan to put up a PR for that, with the caveat that I want to see at least one successful release before we actually remove autotools

Not to pollute this PR too much, but we should also work on fixes for #1078 and #1072 before abandoning autotools

milroy commented:
This PR is close to being mergable. The remaining work is to rebase such that each commit makes one closely related set of changes. @vsoch breaking your last commit into chunks and squashing them into the relevant prior commits as fixups will respect the one-change-per-commit policy but not your significant contributions.

A few rebase and force-pushes ago you had a commit structure that simultaneously preserved the one-change-per-commit policy and also respected your contributions. Can you recreate that structure?

vsoch commented:
> A few rebase and force-pushes ago you had a commit structure that simultaneously preserved the one-change-per-commit policy and also respected your contributions. Can you recreate that structure?

Can you be more specific? The last design that I had (that I think you liked) had one commit that was just me, and I think that's how it is now again? At least I tried to conform to that same structure to not run into the issue again.

vsoch commented:
Looks like I need https://github.com/flux-framework/flux-sched/pull/1080

milroy commented:
> The last design that I had (that I think you liked) had one commit that was just me, and I think that's how it is now again? 

That's true, but beyond improving code Claudia and I wrote now you've also added functionality (the Golang. build support for cmake). Can you break out the cmake contributions into a separate commit ("build: add golang to cmake")?

vsoch commented:
> Can you break out the cmake contributions into a separate commit ("build: add golang to cmake")?

I can try, I think the only way I'd know how is to go back a commit and manually add things back in two groups. It sounds arduous and I'm not sure I can do it, but I'll try.

milroy commented:
You should be able to do an interactive rebase one commit back which will show all the changed files. You can add everything but the `CMakeLists.txt` and `GolangSimple.cmake` as one commit with the same message as the current final commit, and then add the cmake-related files as a separate commit. 

I don't think it will be too bad because there aren't changes from the two updates within the same files.

vsoch commented:
To continue to PR chain! This work is continued in #1081 with the updated commit history!

trws commented:
Thanks for these @grondo. I had the sphinx one on my list bud hadn’t seen the issues yet. I may have a chance to throw up some fixes this morning and this is really helpful.

---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>
On September 27, 2023 at 4:24:53 PM GMT+2, Mark Grondona ***@***.***> wrote:

Yeah, I plan to put up a PR for that, with the caveat that I want to see at least one successful release before we actually remove autotools

Not to pollute this PR too much, but we should also work on fixes for #1078<https://urldefense.us/v3/__https://github.com/flux-framework/flux-sched/issues/1078__;!!G2kpM7uM-TzIFchu!zAsVQTsvN1kjJvKuPyixk18bK5d9B4i74UJ2ZdOKExpBDKM95nXib0e9EsnZcKJEM6GkB13LuYGtSpybkcPcPeCxjVw$> and #1072<https://urldefense.us/v3/__https://github.com/flux-framework/flux-sched/issues/1072__;!!G2kpM7uM-TzIFchu!zAsVQTsvN1kjJvKuPyixk18bK5d9B4i74UJ2ZdOKExpBDKM95nXib0e9EsnZcKJEM6GkB13LuYGtSpybkcPcyWK-aeI$> before abandoning autotools

—
Reply to this email directly, view it on GitHub<https://urldefense.us/v3/__https://github.com/flux-framework/flux-sched/pull/1062*issuecomment-1737509407__;Iw!!G2kpM7uM-TzIFchu!zAsVQTsvN1kjJvKuPyixk18bK5d9B4i74UJ2ZdOKExpBDKM95nXib0e9EsnZcKJEM6GkB13LuYGtSpybkcPcKLIG8PI$>, or unsubscribe<https://urldefense.us/v3/__https://github.com/notifications/unsubscribe-auth/AAFBFNPVOSY6MVANECECXM3X4QZKVANCNFSM6AAAAAA33WZDII__;!!G2kpM7uM-TzIFchu!zAsVQTsvN1kjJvKuPyixk18bK5d9B4i74UJ2ZdOKExpBDKM95nXib0e9EsnZcKJEM6GkB13LuYGtSpybkcPcv058OBU$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

vsoch commented:
> ⚠️ The sha of the head commit of this PR conflicts with https://github.com/flux-framework/flux-sched/pull/1081. Mergify cannot evaluate rules on this PR. ⚠️

For context, we basically took the fixed commits from #1081 and force pushed them here, and that way the one that is finally merged is linked to discussion here. And that PR is closed, so mergify "having a moment" is hopefully resolved!

vsoch commented:
![image](https://github.com/flux-framework/flux-sched/assets/814322/69d9d449-31ce-4601-a48a-55c7abbcb135)

milroy reviewed (approved):
This is an excellent contribution @cmisale and @vsoch. Thank you both for your efforts and patience, and thank you @vsoch for your improvements to the original code and for getting Cmake and Golang to work together!
